File: backend/core-services/product-service/src/main/java/com/movieticket/product/entity/Movie.java
package com.movieticket.product.entity;

import com.movieticket.product.enums.ProjectionType;
import com.movieticket.product.enums.ReleaseStatus;
import com.movieticket.product.enums.TranslationType;
import jakarta.persistence.*;
import lombok.*;
import org.hibernate.annotations.BatchSize;

import java.time.LocalDate;
import java.time.LocalDateTime;
import java.util.Set;

@Entity
@Table(name = "movies")
@Data
@Getter
@Setter
@NoArgsConstructor
@AllArgsConstructor
@Builder
public class Movie {

    @Id
    @GeneratedValue(strategy = GenerationType.UUID)
    private String id;

    private String title;
    private int duration;
    private LocalDate premiereDate;
    private String producer;

    @Column(columnDefinition = "TEXT")
    private String description;

    private String trailerUrl;
    private String thumbnailUrl;
    private String bannerUrl;

    @ElementCollection
    @CollectionTable(name = "movie_projections", joinColumns = @JoinColumn(name = "movie_id"))
    @Enumerated(EnumType.STRING)
    @Column(name = "projection_type")
    private Set<ProjectionType> supportedProjection;

    @ElementCollection
    @CollectionTable(name = "movie_translations", joinColumns = @JoinColumn(name = "movie_id"))
    @Enumerated(EnumType.STRING)
    @Column(name = "translation_type")
    private Set<TranslationType> supportedTranslation;

    @ManyToOne
    @JoinColumn(name = "age_type_id")
    private AgeType ageType;

    private int releaseYear;
    private String nationality;

    @Enumerated(EnumType.STRING)
    private ReleaseStatus releaseStatus = ReleaseStatus.UPCOMING;

    @ElementCollection
    @CollectionTable(name = "movie_genres", joinColumns = @JoinColumn(name = "movie_id"))
    @Column(name = "genre")
    @BatchSize(size = 20) // gom 20 movie lại để truy vấn thể loại của nó cùng 1 lượt
    private Set<String> genre;

    @ManyToMany
    @JoinTable(name = "movie_actors", joinColumns = @JoinColumn(name = "movie_id"), inverseJoinColumns = @JoinColumn(name = "artist_id"))
    private Set<Artist> actors;

    @ManyToMany
    @JoinTable(name = "movie_directors", joinColumns = @JoinColumn(name = "movie_id"), inverseJoinColumns = @JoinColumn(name = "artist_id"))
    private Set<Artist> directors;

    private double avgRating;
    private LocalDateTime createdAt;
    private LocalDateTime updatedAt;

    @PrePersist
    protected void onCreate() {
        createdAt = LocalDateTime.now();
        updatedAt = LocalDateTime.now();
    }

    @PreUpdate
    protected void onUpdate() {
        updatedAt = LocalDateTime.now();
    }
}

File: frontend/web/src/types/movie.ts
import type { ArtistSummary } from "./artist";

export const ReleaseStatus = {
  UPCOMING: "UPCOMING",
  NOW_SHOWING: "NOW_SHOWING",
  ENDED: "ENDED",
};

export type ReleaseStatusType =
  (typeof ReleaseStatus)[keyof typeof ReleaseStatus];

export interface MovieSummaryResponse {
  id: string;
  title: string;
  duration: number;
  premiereDate: Date;
  producer: string;
  thumbnailUrl: string;
  ageType: string;
  releaseYear: number;
  nationality: string;
  avgRating: number;
  releaseStatus: ReleaseStatusType;
  genre: string[];
}

export interface SelectionDTO {
  id: string;
  label: string;
}

export interface MoviePublic {
  id: string;
  title: string;
  duration: number;
  premiereDate: Date;
  thumbnailUrl: string;
  trailerUrl: string;
  bannerUrl: string;
  ageType: string;
  avgRating: number;
}

export interface MoviePublicDetail {
  id: string;
  title: string;
  duration: number;
  description: string;
  premiereDate: Date;
  producer: string;
  thumbnailUrl: string;
  bannerUrl: string;
  ageType: string;
  releaseYear: number;
  nationality: string;
  avgRating: number;
  totalReviews: number;
  releaseStatus: ReleaseStatusType;
  genre: string[];
  actors: ArtistSummary[];
  directors: ArtistSummary[];
  trailerUrl: string;
}

export const ALL_GENRES = [
  "Hành động",
  "Kinh dị",
  "Tình cảm",
  "Hài hước",
  "Viễn tưởng",
  "Phiêu lưu",
];
export const ALL_PROJECTION = ["2D", "3D", "IMAX"];

export const ALL_PROJECTION_OPTIONS = [
  { label: "2D", value: "_2D" },
  { label: "3D", value: "_3D" },
  { label: "IMAX", value: "IMAX" },
];

export const ALL_TRANSLATION = [
  { label: "Lồng tiếng", value: "DUBBING" },
  { label: "Phụ đề", value: "SUBTITLES" },
  { label: "Thuyết minh", value: "VOICE_OVER" },
];
export const ALL_STATUS = ["UPCOMING", "NOW_SHOWING", "ENDED"];

export interface MovieSearchDTO {
  title?: string;
  releaseStatus?: ReleaseStatusType;
  nationality?: string;
  ageTypeId?: string;
  releaseYear?: number;
  genre?: string;
  artistId?: string;
  projectionType?: string;
  translationType?: string;
}
